src/sentry/static/sentry/app/views/settings/components/forms/richListField.jsx

```diff
@@ -0,0 +1,231 @@
+   */
+  addDropdown: PropTypes.shape(DropdownAutoComplete.propTypes).isRequired,
+
+  /**
+   * Render function to render an item.
+   */
+  renderItem: PropTypes.func,
+
+  /**
+   * Callback invoked when an item is added via the dropdown menu.
+   */
+  onAddItem: PropTypes.func,
+
+  /**
+   * Callback invoked when an item is interacted with.
+   */
+  onEditItem: PropTypes.func,
+
+  /**
+   * Callback invoked when an item is removed.
+   */
+  onRemoveItem: PropTypes.func,
+
+  /**
+   * Properties for the confirm remove dialog. If missing, the item will be
+   * removed immediately.
+   */
+  removeConfirm: PropTypes.object,
+};
+
+function getDefinedProps(propTypes, props) {
+  return _.pickBy(props, (_prop, key) => key in propTypes);
+}
+
+class RichList extends React.PureComponent {
+  static propTypes = {
+    ...RichListProps,
+
+    /**
+     * The list of items to render.
+     */
+    value: PropTypes.array.isRequired,
+  };
+
+  static defaultProps = {
+    addButtonText: t('Add Item'),
+    renderItem: item => item,
+    onAddItem: (item, addItem) => addItem(item),
+    onEditItem: () => {},
+    onRemoveItem: (item, removeItem) => removeItem(item),
+  };
+
+  triggerChange = items => {
+    this.props.onChange(items, {});
+    this.props.onBlur(items, {});
+  };
+
+  addItem = data => {
+    const items = [...this.props.value, data];
+    this.triggerChange(items);
+  };
+
+  updateItem = (data, index) => {
+    const items = [...this.props.value];
+    items.splice(index, 1, data);
+    this.triggerChange(items);
+  };
+
+  removeItem = index => {
+    const items = [...this.props.value];
+    items.splice(index, 1);
+    this.triggerChange(items);
+  };
+
+  onSelectDropdownItem = item => {
+    this.props.onAddItem(item, this.addItem);
+  };
+
+  onEditItem = (item, index) => {
+    this.props.onEditItem(item, data => this.updateItem(data, index));
+  };
+
+  onRemoveItem = (item, index) => {
+    this.props.onRemoveItem(item, () => this.removeItem(index));
+  };
+
+  renderItem = (item, index) => {
+    const removeIcon = (
+      <ItemIcon>
+        <InlineSvg src="icon-trash" size="12px" />
+      </ItemIcon>
+    );
+
+    const removeConfirm = this.props.removeConfirm ? (
+      <Confirm
+        priority="danger"
+        confirmText={t('Remove')}
+        {...this.props.removeConfirm}
+        onConfirm={() => this.onRemoveItem(item, index)}
+      >
+        {removeIcon}
+      </Confirm>
+    ) : (
+      React.cloneElement(removeIcon, {
+        onClick: () => this.onRemoveItem(item, index),
+      })
+    );
+
+    return (
+      <Item size="small" key={index}>
+        <ItemLabel>
+          {this.props.renderItem(item)}
+          <ItemIcon onClick={() => this.onEditItem(item, index)}>
+            <InlineSvg src="icon-settings" size="12px" />
+          </ItemIcon>
+          {removeConfirm}
+        </ItemLabel>
+      </Item>
+    );
+  };
+
+  renderDropdown = () => {
+    return (
```

Comment:
Do we know why there's a weird gap in space between the dropdown in the screenshot?

Reply:
Due to the flex container, which is resizing the dropdown's div container. Might go away with switching to grid. Let me play around with this.

Reply:
Fixed in the next commit.